userspace/async/async_metadata_source.tpp

```diff
@@ -0,0 +1,286 @@
+
+	if(!request_complete)
+	{
+		// Haven't made the request yet
+		if (itr == m_metadata_map.end())
+		{
+			m_metadata_map[key].m_available = false;
+			m_metadata_map[key].m_metadata = metadata;
+		}
+
+		// Make request to API and let the async thread know about it
+		if (std::find(m_request_queue.begin(),
+		              m_request_queue.end(),
+		              key) == m_request_queue.end())
+		{
+			m_request_queue.push_back(key);
+			m_queue_not_empty_condition.notify_one();
+		}
+
+		//
+		// If the client code is willing to wait a short amount of time
+		// to satisfy the request, then wait for the async thread to
+		// pick up the newly-added request and execute it.  If
+		// processing that request takes too much time, then we'll
+		// not be able to return the metadata information on this call,
+		// and the async thread will continue handling the request so
+		// that it'll be available on the next call.
+		//
+		if (m_max_wait_ms > 0)
+		{
+			m_metadata_map[key].m_available_condition.wait_for(
+					guard,
+					std::chrono::milliseconds(m_max_wait_ms));
+
+			itr = m_metadata_map.find(key);
+			request_complete = (itr != m_metadata_map.end()) && itr->second.m_available;
+		}
+	}
+
+	if(request_complete)
+	{
+		metadata = itr->second.m_metadata;
+		m_metadata_map.erase(key);
+	}
+	else
+	{
+		m_metadata_map[key].m_callback = callback;
+	}
+
+	return request_complete;
+}
+
+template<typename key_type, typename metadata_type>
+std::size_t async_metadata_source<key_type, metadata_type>::queue_size() const
+{
+	std::lock_guard<std::mutex> guard(m_mutex);
+	return m_request_queue.size();
+}
+
+template<typename key_type, typename metadata_type>
+key_type async_metadata_source<key_type, metadata_type>::dequeue_next_key()
+{
+	std::lock_guard<std::mutex> guard(m_mutex);
+	key_type key = m_request_queue.front();
+
+	m_request_queue.pop_front();
+
+	return key;
+}
+
+template<typename key_type, typename metadata_type>
+metadata_type async_metadata_source<key_type, metadata_type>::get_metadata(
+		const key_type& key)
+{
+	std::lock_guard<std::mutex> guard(m_mutex);
+
+	return m_metadata_map[key].m_metadata;
+}
+
+template<typename key_type, typename metadata_type>
+void async_metadata_source<key_type, metadata_type>::store_metadata(
+		const key_type& key,
+		const metadata_type& metadata)
+{
+	std::lock_guard<std::mutex> guard(m_mutex);
+
+	if (m_metadata_map[key].m_callback)
+	{
+		m_metadata_map[key].m_callback(key, metadata);
+		m_metadata_map.erase(key);
+	}
+	else
+	{
+		m_metadata_map[key].m_metadata = metadata;
+		m_metadata_map[key].m_available = true;
+		m_metadata_map[key].m_available_condition.notify_one();
+	}
+}
+
+/**
+ * Prune any "old" outstanding requests.  This method expect that the caller
+ * is holding m_mutex.
+ */
+template<typename key_type, typename metadata_type>
+void async_metadata_source<key_type, metadata_type>::prune_stale_requests()
```

Comment:
the assumption is we have a small enough number of requests that this is not going to be an issue?

Reply:
There's that assumption, yes, and this also happens in the context of the async thread.